API/API.mli

```diff
@@ -4645,16 +4645,23 @@ sig
   type proof
   type 'a focus_kind
 
+  val proof : proof ->
+    Goal.goal list * (Goal.goal list * Goal.goal list) list *
+    Goal.goal list * Goal.goal list * Evd.evar_map
+
   val run_tactic : Environ.env ->
                    unit Proofview.tactic -> proof -> proof * (bool * Proofview_monad.Info.tree)
   val unshelve : proof -> proof
   val maximal_unfocus : 'a focus_kind -> proof -> proof
   val pr_proof : proof -> Pp.t
+
   module V82 :
   sig
     val grab_evars : proof -> proof
 
     val subgoals : proof -> Goal.goal list Evd.sigma
+    [@@ocaml.deprecated "Use the first and fifth argument of [Proof.proof]"]
```

Comment:
I wouldn't say argument, I would say component of returned tuple.
However, I also find returning such a big tuple to be a very bad idea, why not a record instead.

Reply:
IMHO the whole API needs a revision, but it shouldn't be done piece-wise, but in a more comprehensive way and hopefully at once.

Compatibility is the problem here, as we cannot [shouldn't] modify this API without deprecating it first, so the revised API should come as a module.


> why not a record instead.

That's what the V82 API did.

Reply:
I'm saying this because I'm not sure that it makes sense to deprecate a function in favor of an alternative that we find less good, especially if it's to deprecate the alternative function soon.
We should rather work on the new alternative which satisfies us, and then deprecate all the previous revisions at once.

Reply:
Umm, I think the function was already "deprecated" long time ago, I think. We could then "undeprecate" such functions and get them out of the V8.2 namespace, but something has to be done.

Again the intention of this patch is to consolidate the codebase [like quite a few other] rather than trying to improve previous choices. For example, the "Feedback" system is far from optimal in my opinion, but once it was introduced the only step (other than reverting it) is to consolidate everything, then we can hope to improve it.

Reply:
We should agree on the definition of deprecating. To me it means emitting a warning so that the user can fix in advance a breakage that is going to occur in the next version when we'll remove the function.

That being said I said what I had to say, I'm not going to argue more nor am I going to oppose to this PR.

Reply:
Interesting; we could indeed deprecate `Proof.proof` except that we don't have a replacement. I could introduce one, but however I do think that a deeper evolution on this part of the API will be needed so maybe it is best to wait.